### DISCUSSION TITLE
SSH Private Key with Renovate On Prem

### DISCUSSION THREAD

**Discussion by chrispoupart** [renovatebot/renovate#14100] [Request Help]
Hey folks.

I've been trying to figure this one out for a little while, and I feel like I'm missing something obvious.

I've got a whole bunch of private Ansible roles stored in private GitLab repositories. All the projects are configured with the same SSH deploy key, meaning that using the private ssh key will let me clone any one of the projects.

So, with that configured properly, I can use `ansible-galaxy` to install all of the roles in my `requirements.yml` file.  This works well from our CI.

The problem comes with trying to configure Renovate On-Prem to use this ssh key.

By default, even though the Renovate Bot account should have access to the projects (in fact, if I impersonate the bot in GitLab, I can access and clone the projects, over https), it seems unable to check them and returns the following in the onboarding merge request:

### ⚠ Dependency Lookup Warnings ⚠

Please correct - or verify that you can safely ignore - these lookup failures before you merge this MR.

-   `Failed to look up dependency ansible_history`

## Attempted Configurations

In order to run Renovate On-Prem, I'm using Podman. To inject the SSH key into the container, I'm using [podman secrets](https://docs.podman.io/en/latest/markdown/podman-secret-create.1.html), which basically allow you to securely mount a secret. However you can't control the mounted location, and they end up in the container under `/var/run/secrets/<secret name>`.

So... to tell git to use the key located in `/var/run/secrets/role-deploy-key`, I also configured the container environment variable `GIT_SSH_COMMAND=/usr/bin/ssh -i /var/run/secrets/role-deploy-key`. (Basically, changing the default identity path for any SSH command). If I enter the container, I can validate that variable is properly set, the key is in place, and I can manually clone one of the git projects over ssh without any problems.

However, that doesn't seem to be working for the Renovate Bot.

If, on the other hand, I create the key under `~/.ssh/id_rsa` with the proper permissions/mode, then the Renovate Bot does what it is supposed to do and can properly check the roles listed in my requirements.yml file.

Unfortunately... short of writing some sort of entrypoint script that messes around with this on startup and building my own version of the container, I'm not sure how exactly I'm supposed to do this.  Obviously, we don't want to build an image (even for internal use) that has the secret baked in, though I suppose I could try adding an ~/.ssh/config that points to the mounted secret...

In case it matters, my `requirements.yml` file follows the standard:

```yaml
roles:
  - src: git@gitlab.example.com:ansible/roles/role-ansible_history.git
    name: ansible_history
    version: 1.4.3
    path: roles/
    scm: git
```

Thanks for any pointers in the right direction!

**Comment by chrispoupart** [comment]
While I wait to see if there are any better solutions, I've done the following and it works as expected:

1. Built my own container to add an SSH config file:
    ```
    FROM registry.hub.docker.com/whitesource/renovate:2.1.1
    COPY --chmod=0644 --chown=1000:1000 ssh_config /home/ubuntu/.ssh/config
    ```
1. The `ssh_config` file contained a config with the following:
    ```
   Host *
      IdentityFile  /var/run/secrets/role-deploy-key
    ```
1. As mentioned above, I inject the key into the container at runtime using the Podman secrets feature.

That seems to work, but will require additional overhead to build our own containers (not the end of the world, was just hoping it wouldn't be necessary).

This then meets the requirements of making the SSH key available to Renovate without baking it into the image.

**Reply by viceice** [reply]
You should be able to mount those files via volume at runtime, so no need for custom image

**Reply by chrispoupart** [reply]
This is true, and I probably should have mentioned that that is not our first choice for a few reasons (described below).

TBH, a custom image is likely going to be on the docket anyway, as it seems that while the default user is "ubuntu", it is also a member of the "root" group, and our security policies explicitly don't allow root-permission container images (and while podman is currently running this POC as root, the plan is for it to be run as a non-root user).  This also comes into play mounting volumes in a root-less container, because the "ubuntu" user in the container won't have permission to write to files like `known_hosts` as "ubuntu:root" (which is what it does now, but would have access as "ubuntu:ubuntu" (assuming a uid:gid of 1000:1000 and that the host user that will run podman has a matching uid:gid).  That, and the fact that the container has a multitude of (mostly low and moderate) security issues (by way of, ironically, outdated packages).  So in building on the existing container to tighten the security posture of the user/group settings, and updating the container os packages, we may as well just drop in a custom ssh config!

The other aspect comes down to security best practices.  While you can mount the volume in at runtime, it's not the best option for secrets as your risk leaking them if you do anything like `podman commit` or `podman export` (whereas podman secrets are explicitly designed not to be included in any container exports).

Most of my recent container work has been with OpenShift and GKE. Normally this just would have been done with a [Secret](https://kubernetes.io/docs/concepts/configuration/secret/) configuration mounted as a volume. I was disappointed that Podman secrets don't work the same way.

**Reply by rarkins** [reply]
Root group is not the same as root user. I think either your security team has a misunderstanding or we do. It's built this way for OpenShift reasons, are you familiar that Openshift uses "root group" too?

**Reply by chrispoupart** [reply]
No, the misunderstanding was definitely a brain fart on my side! Getting volumes mounted in root-less podman instances is not obvious and I was misattributing my conflicts to the gid=0 (it wasn't).

Finally what I'm going to be going with is a mix of the two suggestions:

Mounting the volume into `/home/ubuntu/.ssh` (instead of building my own container) but not the SSH key. Instead it will include a config that points to the key added to the container via `podman --secret`.

Thanks very much for your help!

**Reply by chrispoupart** [reply]
So... this didn't work properly for me.

Does renovate respect the `~/.ssh/config` settings? Or does it simply look in `~/.ssh/` for appropriate keys?

Setup as described, I can enter the container and manually git clone a project, but Renovate is still complaining about not being able to access them. :(

**Reply by rarkins** [reply]
Renovate looks up dependencies over https. What does the datasource and depName/lookupName look like for this dependency?

**Reply by chrispoupart** [reply]
So I haven't configured any rules beyond the defaults at this point.  Sometimes in my mucking around things have worked as expected. Other times not and the onboarding MR gets updated with a warning that the Ansible Galaxy dependencies can not be verified.

I've been doing a lot of configuration changes as I build up an Ansible playbook to deploy this fully, so I'm no longer sure under what conditions it succeeds and under which conditions it fails. That being said, most of the changes that impact the container configuration have been related to the SSH keys mentioned here, which shouldn't matter if the dependency lookup is over https :-/.

I'll do some more thorough testing tomorrow so I don't waste any more of your time.

**Reply by chrispoupart** [reply]
OK, so I think that I figured it out.

It _is_ using ssh to validate the Ansible roles and collections referenced as git repos in the requirements.yml file.  I don't know if that is intended behaviour or not.  

What was preventing my above configuration from working was SSH TOFU (trust on first use).  If I pre-populate the ~/.ssh/known_hosts value with my GitLab server keys, then the above-presented configuration works as expected.

**Reply by viceice** [reply]
ok, can you summarize all required steps in a new answer? We can then maybe add that to our docs as sample / faq. 🤗

**Reply by chrispoupart** [reply]
OK, sorry for the delay on this. I had to get my configuration merged and deployed to test things out before I wrote it up.

So, the situation is this (I can also provide a sample Ansible playbook for this, if there is interest):

## Problem

1. We have Ansible roles and collections mirrored in our on-prem GitLab server (this could be any git repository).
2. Our Ansible projects have a `requirements.yml` file [configured to install roles/collections from git](https://galaxy.ansible.com/docs/using/installing.html#installing-multiple-roles-from-a-file), and more importantly, configured to access the git repositories using git over SSH.  This access requires the presence of a private SSH key.

How do we get the SSH key into the Renovate On-Prem container in a secure fashion?

## Solution

There are many different potential solutions here, and they will depend largely on what container tools you are using. Obviously, Kubernetes will be different from Docker, which will be different from Podman, etc.

In our case, we are using Podman, and we are making our secret available in Podman using [Podman secrets](https://docs.podman.io/en/latest/markdown/podman-secret.1.html).  Secrets mounted in Podman this way are made available inside the container at `/var/run/secrets/<secret-name>`.  So by itself, this is insufficient for ssh from within the container to access the git repositories (as it doesn't know where to look for the key).

So the next part of the solution is to create a volume containing the required SSH configurations, and mount it under `/home/ubuntu/.ssh`. The volume must contain 2 files:

1. `config` which contains the following (where `role-deploy-key` is the name of the secret we used):
   ```
   Host *
     IdentifyFile /var/run/secrets/role-deploy-key
   ```
2. `known_hosts`, populated with the server keys of our GitLab server, so that we did not have to worry about SSH "trust on first use".

With these settings, managed all via Ansible, we are able to spin up a container that can access our private internal Ansible roles and collections using git over ssh. This also allows Renovate to properly check versions for said roles and collections (which are by convention git tags in semantic versioning format).

**Reply by faenschi** [reply]
hello @chrispoupart 

i'm also trying to configure renovate to use a ssh key. for testing purposes i build the docker image with the ssh-config file and mount my key when I start the container (later this will run on kubernetes with a mounted volume). 

I start the container with the following options:

```
docker run \
--rm \
-e LOG_LEVEL="TRACE" \
-e GITHUB_COM_TOKEN="$GITHUB_COM_TOKEN" \
-e RENOVATE_PLATFORM="gitlab" \
-e RENOVATE_GIT_URL="ssh" \
-e RENOVATE_LOG_FILE_LEVEL="trace" \
-e RENOVATE_REPOSITORIES="myrepo" \
-v "/export/home/blabla/.ssh/id_rsa:/home/ubuntu/.ssh/id_rsa" \
renovate:latest
```
but then i get the following error:  `You must configure a GitLab personal access token`

But this is exactly what i don't want, to use an access token with renovate. How did you manage to start the renovate-bot?

kind regards 
angela

**Reply by viceice** [reply]
renovate needs the token to access the gitlab API for issue and PR management. there you can't use ssh.

**Reply by faenschi** [reply]
Thanks, i saw it now as well. so there is no possibility to just use ssh i guess. A token is always needed to access the api...

**Comment by jonesbusy** [comment]
Found this thread because faced the same issue trying to update a private Galaxy collection reference on private Bitbucket Server.

```
"message": "Host key verification failed.\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n",
```

I'm running renovate on a Jenkins pipeline (Kubernetes container).

My `requirements.yml`

```
collections:
  - name: ssh://git@<bitbucket_host>:7999/<project>/<repo>.git#/<path>/
    version: 1.0.1
    type: git
```

All usual CLI operation on git works with GIT_SSH_COMMAND but is totally ignored by renovate git-tags datasource.

Only solution I found to make it work is to copy a `~/.ssh/config` and `~/.ssh/id_rsa` on the current container.

I don't really like this solution but I didn't find better
